### PR TITLE
s390x build support

### PR DESCRIPTION
**Pull request by suzyw-w**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
N/A

Support build on s390x platform.

This PR depends on #45519.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -453,7 +453,7 @@ endif ()
 set (CMAKE_POSTFIX_VARIABLE "CMAKE_${CMAKE_BUILD_TYPE_UC}_POSTFIX")
 
 set (CMAKE_POSITION_INDEPENDENT_CODE OFF)
-if (OS_LINUX AND NOT ARCH_AARCH64)
+if (OS_LINUX AND NOT ARCH_AARCH64 AND NOT ARCH_S390X)
```
Comment (rschu1ze) [inline CMakeLists.txt:456]: I am rather confused by l. 456-462. Before your change, PIE was disabled for x86, PPC and s390x (i.e. `NOT ARCH_AARCH64`, l. 456). But the comment says PIE is "disabled" on ARM - actually ARM is the only platform where PIE is enabled. While at it, would you like to change the wording of the comment here? Thanks.

Please also add a comment along the lines of l. 458 explaining *why* PIE is disabled as well on s390x.

File: CMakeLists.txt [hunk 2]
```diff
@@ -471,7 +471,11 @@ enable_testing() # Enable for tests without binary
 
 option(ENABLE_OPENSSL "This option performs a build with OpenSSL. NOTE! This option is insecure and should never be used. By default, ClickHouse uses and only supports BoringSSL" OFF)
 
-option(ENABLE_OPENSSL_DYNAMIC "This option removes SSL from ClickHouse and will link to the OpenSSL version supplied by OS." OFF)
+if(ARCH_S390X)
```
Comment (rschu1ze) [inline CMakeLists.txt:474]: Instead of l. 474, please use something like:
```cmake
if (ARCH_S390X)
    set(ENABLE_OPENSSL_DYNAMIC_DEFAULT, ON)
else ()
    set(ENABLE_OPENSSL_DYNAMIC_DEFAULT, OFF)
endif ()
option(ENABLE_OPENSSL_DYNAMIC "This option removes SSL from ClickHouse and will link to the OpenSSL version supplied by OS." ${ENABLE_OPENSSL_DYNAMIC_DEFAULT})
```

File: PreLoad.cmake [hunk 3]
```diff
@@ -58,10 +58,9 @@ execute_process(COMMAND uname -m OUTPUT_VARIABLE ARCH)
 # By default, prefer clang on Linux
 # But note, that you still may change the compiler with -DCMAKE_C_COMPILER/-DCMAKE_CXX_COMPILER.
 if (OS MATCHES "Linux"
+    # some build systems may use CC/CXX env variables
     AND "$ENV{CC}" STREQUAL ""
-    AND "$ENV{CXX}" STREQUAL ""
-    AND NOT DEFINED CMAKE_C_COMPILER
-    AND NOT DEFINED CMAKE_CXX_COMPILER)
+    AND "$ENV{CXX}" STREQUAL "")
```
Comment (rschu1ze) [inline PreLoad.cmake:63]: Maybe I missed it but how is the removal of l. 63/64 (left side) related to s390/x?

File: base/base/defines.h [hunk 4]
```diff
@@ -28,7 +28,7 @@
 #define NO_INLINE __attribute__((__noinline__))
 #define MAY_ALIAS __attribute__((__may_alias__))
 
-#if !defined(__x86_64__) && !defined(__aarch64__) && !defined(__PPC__) && !(defined(__riscv) && (__riscv_xlen == 64))
+#if !defined(__x86_64__) && !defined(__aarch64__) && !defined(__PPC__) && !defined(__s390x__) && !(defined(__riscv) && (__riscv_xlen == 64))
```
Comment (rschu1ze) [inline base/base/defines.h:31]: Please also adjust l. 32

File: cmake/arch.cmake [hunk 5]
```diff
@@ -7,8 +7,12 @@ elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64.*|AARCH64.*|arm64.*|ARM64.*)")
     set (ARCH_AARCH64 1)
 elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc64le.*|ppc64le.*|PPC64LE.*)")
     set (ARCH_PPC64LE 1)
+elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(s390x.*|S390X.*)")
+    set (ARCH_S390X 1)
 elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "riscv64")
     set (ARCH_RISCV64 1)
+elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "^(s390x.*|S390X.*)")
```
Comment (rschu1ze) [inline cmake/arch.cmake:14]: Duplicates l. 10

File: cmake/linux/toolchain-s390x.cmake [hunk 6]
```diff
@@ -0,0 +1,25 @@
+set (CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
```
Comment (rschu1ze) [inline cmake/linux/toolchain-s390x.cmake:1]: Please add an `include_guard` - see the other toolchain files in "cmake/linux/" for examples.

File: src/Compression/LZ4_decompress_faster.cpp [hunk 7]
```diff
@@ -81,7 +81,7 @@ inline void copyOverlap8(UInt8 * op, const UInt8 *& match, size_t offset)
 }
 
 
-#if defined(__x86_64__) || defined(__PPC__) || defined(__riscv)
+#if defined(__x86_64__) || defined(__PPC__) || defined(__riscv) || defined(__s390x__)
```
Comment (rschu1ze) [inline src/Compression/LZ4_decompress_faster.cpp:84]: Cosmetic: could we keep the sort order used elsewhere: x86, ARM, PPC, s390x, riscv?

File: src/Storages/StorageS3Cluster.cpp [hunk 8]
```diff
@@ -10,6 +10,7 @@
 #include <DataTypes/DataTypeString.h>
 #include <IO/WriteBufferFromS3.h>
 #include <IO/WriteHelpers.h>
+#include <IO/ConnectionTimeoutsContext.h>
```
Comment (rschu1ze) [inline src/Storages/StorageS3Cluster.cpp:13]: #include seems unrelated

File: src/Storages/checkAndGetLiteralArgument.cpp [hunk 9]
```diff
@@ -36,5 +36,5 @@ template UInt64 checkAndGetLiteralArgument(const ASTPtr &, const String &);
 template UInt8 checkAndGetLiteralArgument(const ASTPtr &, const String &);
 template bool checkAndGetLiteralArgument(const ASTPtr &, const String &);
 template String checkAndGetLiteralArgument(const ASTLiteral &, const String &);
-
+template UInt64 checkAndGetLiteralArgument(const ASTLiteral &, const String &);
```
Comment (rschu1ze) [inline src/Storages/checkAndGetLiteralArgument.cpp:39]: Looks unrelated

File: src/Common/formatIPv6.cpp [hunk 10]
```diff
@@ -165,9 +165,9 @@ void formatIPv6(const unsigned char * src, char *& dst, uint8_t zeroed_tail_byte
             uint8_t ipv4_buffer[IPV4_BINARY_LENGTH] = {0};
             memcpy(ipv4_buffer, src + 12, IPV4_BINARY_LENGTH);
             // Due to historical reasons formatIPv4() takes ipv4 in BE format, but inside ipv6 we store it in LE-format.
-            if constexpr (std::endian::native == std::endian::little)
-                std::reverse(std::begin(ipv4_buffer), std::end(ipv4_buffer));
-
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Comment (rschu1ze) [inline src/Common/formatIPv6.cpp:168]: Unless there are good reasons for the C-style macro form, I'd prefer the C++-native `std::endian`.

File: cmake/tools.cmake [hunk 11]
```diff
@@ -53,20 +53,23 @@ list (GET COMPILER_VERSION_LIST 0 COMPILER_VERSION_MAJOR)
...
-        find_program (LLD_PATH NAMES "ld.lld")
-        find_program (GOLD_PATH NAMES "ld.gold")
-    elseif (COMPILER_CLANG)
-        # llvm lld is a generic driver.
-        # Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
-        if (OS_LINUX)
-            find_program (LLD_PATH NAMES "ld.lld-${COMPILER_VERSION_MAJOR}" "ld.lld")
-        elseif (OS_DARWIN)
-            find_program (LLD_PATH NAMES "ld64.lld-${COMPILER_VERSION_MAJOR}" "ld64.lld")
+# s390x doesnt support lld
```
Comment (rschu1ze) [inline cmake/tools.cmake:56]: Okay, so the heuristics (l. 58ff) that picks a linker if none was explicitly specified (l. 54) doesn't work on s390x. I think we should rather assert that the user provided a linker explicitly. What about something like this:

```cmake
option (LINKER_NAME "Linker name or full path")

if (ARCH_S390X AND NOT LINKER_NAME)
    message (FATAL_ERROR "On s390x, the linker must be specified manually.")
endif ()

if (NOT LINKER_NAME)
    ...
```

Comment (SuzyWangIBMer) [reply cmake/tools.cmake:56]: Since s390x doesnt support lld, it doesnt matter if linker_name is there or not. lld path should not be included in s390x case.

File: src/Common/formatIPv6.cpp (hunk 10 above)
Comment (SuzyWangIBMer) [reply src/Common/formatIPv6.cpp:168]: native style doesn't compile on s390x platform. error: code will never be executed [-Werror,-Wunreachable-code]

Comment (rschu1ze) [reply src/Common/formatIPv6.cpp:168]: Hm, okay. Perhaps an issue with the compiler on s390x.

File: base/base/defines.h (hunk 4 above)
Comment (SuzyWangIBMer) [reply base/base/defines.h:31]: Updated as suggested

File: cmake/linux/toolchain-s390x.cmake (hunk 6 above)
Comment (SuzyWangIBMer) [reply cmake/linux/toolchain-s390x.cmake:4]: Updated as suggested

File: src/Storages/StorageS3Cluster.cpp (hunk 8 above)
Comment (SuzyWangIBMer) [reply src/Storages/StorageS3Cluster.cpp:13]: Compile error on s390x. Seems missing header file for function getTCPTimeoutsWithFailover(), not sure why this only appears on s390x. 
```./src/Storages/StorageS3Cluster.cpp:130: undefined reference to `DB::ConnectionTimeouts::getTCPTimeoutsWithFailover(DB::Settings const&)' ```

File: src/Storages/checkAndGetLiteralArgument.cpp (hunk 9 above)
Comment (SuzyWangIBMer) [reply src/Storages/checkAndGetLiteralArgument.cpp:39]: Also compile error on s390x, invalid conversion on this following two lines.

```./build/./src/TableFunctions/TableFunctionGenerateRandom.cpp:66: undefined reference to `unsigned long DB::checkAndGetLiteralArgument<unsigned long>(DB::ASTLiteral const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'
/usr/bin/s390x-linux-gnu-ld: src/libdbms.a(StorageGenerateRandom.cpp.o): in function `operator()':
./build/./src/Storages/StorageGenerateRandom.cpp:477: undefined reference to `unsigned long DB::checkAndGetLiteralArgument<unsigned long>(DB::ASTLiteral const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)'```

File: CMakeLists.txt [hunk 12]
```diff
@@ -471,7 +471,12 @@ enable_testing() # Enable for tests without binary
 
 option(ENABLE_OPENSSL "This option performs a build with OpenSSL. NOTE! This option is insecure and should never be used. By default, ClickHouse uses and only supports BoringSSL" OFF)
 
-option(ENABLE_OPENSSL_DYNAMIC "This option removes SSL from ClickHouse and will link to the OpenSSL version supplied by OS." OFF)
+if (ARCH_S390X)
+    set(ENABLE_OPENSSL_DYNAMIC_DEFAULT, ON)
```
Comment (rschu1ze) [inline CMakeLists.txt:475]: No commas in `set`, please. (yes, I know I suggested using `set` in a prior comment and I accidentally added a comma into my example)

File: src/Storages/StorageS3Cluster.cpp (hunk 8 above)
Comment (rschu1ze) [reply src/Storages/StorageS3Cluster.cpp:13]: Somehow the style check complains that this #include is duplicated: https://s3.amazonaws.com/clickhouse-test-reports/45520/dea7e2e1f4d5700efb1a339ede45b17c837c2c1b/style_check/duplicate_includes_output.txt

(because of that, the "merge" button is also greyed out for me)

I think we need to get rid of the include. What I don't understand is if you get a compiler error or a link error on s390x. The error you pasted is a link error for sure.

What I see is that function `getTCPTimeoutsWithFailover()` is declared in IO/ConnectionTimeouts.h and defined in IO/ConnectionTimeoutsContext.h, as an `inline` function in a header file though. This means that depending on whether
- the linker actually inlined the code, it will work (and that seems to be the case in ClickHouse CI) or 
- the linker did not inline the code, it will work only if StorageS3Cluster.cpp's library links the library which contains IO/ConnectionTimeoutsContext.h.

This is a general issue which I will try to fix separately.

As a workaround, you could try if replacing `inline` by `ALWAYS_INLINE` in IO/ConnectionTimeoutsContext.h will help.

File: src/Storages/checkAndGetLiteralArgument.cpp (hunk 9 above)
Comment (rschu1ze) [reply src/Storages/checkAndGetLiteralArgument.cpp:39]: Hm, okay

File: src/Storages/StorageS3Cluster.cpp (hunk 8 above)
Comment (SuzyWangIBMer) [reply src/Storages/StorageS3Cluster.cpp:13]: After using `ALWAY_INLINE` instead of `inline`, got multiple definition error. I believe this is a linker error. This will probably take a while to debug.

> multiple definition of DB::ConnectionTimeouts::getTCPTimeoutsWithFailover(DB::Settings const&); src/Functions/CMakeFiles/clickhouse_functions_obj.dir/./catboostEvaluate.cpp.o:catboostEvaluate.cpp:(.text+0x0): first defined here


Can we do something like the following just to make sure it builds for now?
```suggestion
#if defined(__s390x__)
#include <IO/ConnectionTimeoutsContext.h>
#endif
```

Comment (rschu1ze) [reply src/Storages/StorageS3Cluster.cpp:13]: That's okay.

Alternatively, wait for #46006 which will fix the underlying problem.

File: cmake/tools.cmake (hunk 11 above)
Comment (bkuschel) [reply cmake/tools.cmake:56]: For s390x the gnu linker will be used by default, so they should not have to supply the linker name. The preference for the LLVM linker (when it's installed) should be overridden in the s390x case.

File: src/Storages/StorageS3Cluster.cpp (hunk 8 above)
Comment (rschu1ze) [reply src/Storages/StorageS3Cluster.cpp:14]: Since you merged master into this PR and, as a result, got #46006, you would need to remove `#include <IO/ConnectionTimeoutsContext.h>` from this file. ConnectionTimeoutsContext was removed in that other PR.

(I obviously don't have permissions to push to repo clibmouse, otherwise I would have done this little change by myself)

Comment (SuzyWangIBMer) [reply src/Storages/StorageS3Cluster.cpp:14]:
```suggestion
```

**Comment by rschu1ze** [comment]
This PR is generally ready-for-merge (thanks) but the merge button is greyed out for me because there is a merge conflict. Could you please resolve it first (i.e. merge upstream master into your fork)? Thanks!

**Comment by rschu1ze** [comment]
ClickHouse Integration Tests (asan) [4/6]:
ClickHouse Integration Tests (tsan) [4/6]:
- test_storage_rabbitmq/test.py::test_rabbitmq_address #45160 

ClickHouse Integration Tests (release) [4/4]
- test_cluster_copier/test.py::test_copy_month_to_week_partition_with_recovering #30399 

ClickHouse Stateless Tests (release, Databasereplicated) [4/4]:
- 01304_direct_io_long: infra timeout
